Parser/Token.cs

```diff
@@ -38,9 +39,20 @@ public enum E {
    // Utility function used to echo an error to the console
    public void PrintError () {
       if (Kind != ERROR) throw new Exception ("PrintError called on a non-error token");
-      Console.ForegroundColor = ConsoleColor.Yellow;
-      Console.WriteLine ($"At line {Line}, column {Column} of {Source.FileName}: {Text}");
-      Console.ResetColor ();
+      string s = $"File: {Source.FileName}", msg = $"{s}\r\n{new string ('\u2500', s.Length)}\r\n";
+      var (idx, n, space) = (Line - 2, Source.Lines.Length, new string (" "));
+      for (int i = idx; i <= Line; i++, idx++)
+         if (idx >= 1) msg += $"{idx,4}\u2502{space}{Source.Lines[idx - 1]}\r\n";

```

Comment:
There are 2 for loops here - one to print the lines above the error line, and one to print the lines below the error line. Try to implement this with a single for loop. The case where i == Line is a special case where after printing the line you will print also the error text.